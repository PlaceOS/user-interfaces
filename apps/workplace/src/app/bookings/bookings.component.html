<ng-container [ngSwitch]="flow">
    <app-desk-flow (menu)="show_menu = $event" *ngSwitchCase="'desks-map'"></app-desk-flow>
    <desk-flow-listing (menu)="show_menu = $event" *ngSwitchCase="'desks'"></desk-flow-listing>
    <space-flow
        (menu)="show_menu = $event"
        *ngSwitchDefault
    ></space-flow>
    <qr-reader
        *ngSwitchCase="'qr'"
        (menu)="show_menu = $event"
    >
    </qr-reader>

</ng-container>
<a-overlay-menu [(show)]="show_menu"></a-overlay-menu>
