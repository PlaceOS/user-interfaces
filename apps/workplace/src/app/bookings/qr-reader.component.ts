import { Component, OnInit, Output, EventEmitter, ViewChild, ElementRef, OnDestroy } from '@angular/core';
import { Router } from '@angular/router';
import { BrowserQRCodeReader } from '@zxing/library';
import { BaseClass } from '@placeos/common';

@Component({
    selector: 'qr-reader',
    template: `
        <header class="header">
            <a-topbar-header (menuChange)="menu.emit($event)"></a-topbar-header>
        </header>
        <div
            class="flex-1 overflow-hidden flex items-center justify-center bg-black relative"
        >
            <video
                class="min-w-full min-h-full object-cover"
                id="video"
            ></video>
            <div
                class="absolute text-white text-center inset-0 flex flex-col items-center justify-center"
            >
                <div
                    class="flex flex-col z-10 justify-end items-center relative"
                    *ngIf="is_scanning"
                >
                    <h2 class="text-3xl mb-1 uppercase subpixel-antialiased">
                        Scan QR Code
                    </h2>
                    <span class=" mb-4 ">
                        Scan the QR code outisde a PlaceOS room or space.
                    </span>
                </div>
                <div
                    class="flex flex-col z-10 justify-end items-center relative"
                    *ngIf="!is_scanning"
                >
                    <h2 class="text-3xl mb-1 uppercase subpixel-antialiased">
                        Enter Room ID
                    </h2>
                    <span class="mb-4">
                        Enter the room ID number outisde a PlaceOS room or
                        space.
                    </span>
                </div>
                <div class="flex items-center justify-center">
                    <div
                        box
                        class="rounded-2xl h-64 w-64 transition-all flex items-center justify-center p-8 m-8 space-x-2"
                        [class.input]="!is_scanning"
                    >
                        <span class="uppercase">Booking ID</span>
                        <input
                            matInput
                            class="border-none bg-none w-full text-3xl text-left"
                            [(ngModel)]="room_code"
                            name="booking-id"
                            placeholder="e.g. 12102910"
                        />
                    </div>
                </div>
                <div
                    class="m-4 p-2 flex items-center space-x-2 bg-white bg-opacity-50 rounded"
                >
                    <button
                        mat-button
                        [class]="
                            'flex-1 text-black border-none w-40 ' +
                            (is_scanning
                                ? 'bg-white'
                                : 'hover:bg-white bg-opacity-50')
                        "
                        (click)="is_scanning = true"
                    >
                        Scan Code
                    </button>
                    <button
                        mat-button
                        [class]="
                            'flex-1 text-black border-none w-40 ' +
                            (!is_scanning
                                ? 'bg-white'
                                : 'hover:bg-white bg-opacity-50')
                        "
                        (click)="is_scanning = false"
                    >
                        Enter Code
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [
        `
            :host {
                width: 100%;
                height: 100vh;
                display: flex;
                flex-direction: column;
            }

            h2 {
                font-family: 'Grotesk Condensed';
            }

            [box] {
                box-shadow: 0px 0px 0px 100vw rgba(0, 0, 0, 0.5);
            }

            [box] > * {
                display: none;
            }

            [box].input {
                width: 32rem !important;
                max-width: calc(100% - 2rem) !important;
                padding: 1rem !important;
                height: 4rem !important;
                color: black !important;
                background: white;
                box-shadow: 0px 0px 0px 100vw rgba(0, 0, 0, 0.8);
            }

            [box].input > * {
                display: initial;
            }

            [box] span {
                font-family: 'Grotesk Bold';
                text-transform: uppercase;
                letter-spacing: 0.05em;
            }

            button {
                background: transparent;
            }
        `,
    ],
})
export class QrReaderComponent extends BaseClass implements OnInit, OnDestroy {
    /** Whether to show the overlay menu */
    public show_menu: boolean;
    /** Boolean to toggle scan/code */
    public is_scanning: boolean = true;
    /** Room Code input value */
    public room_code: string;
    /** Menu event */
    @Output() public menu = new EventEmitter(false);
    /** Device ID of Camera */
    private device_id: string;
    /** Video element to emit camera feed */
    @ViewChild('video') private _video_el: ElementRef<HTMLVideoElement>;

    constructor(private _router: Router) {
        super();
    }

    public ngOnDestroy() {
        if (this._video_el?.nativeElement.srcObject) {
            const track = (this._video_el.nativeElement
                .srcObject as any).getTracks()[0];
            if (track) {
                track.stop();
            }
        }
    }

    public ngOnInit(): void {
        const codeReader = new BrowserQRCodeReader();
        codeReader.listVideoInputDevices().then((videoInputDevices) => {
            this.device_id = videoInputDevices[0].deviceId;
            codeReader
                .decodeOnceFromVideoDevice(this.device_id, 'video')
                .then((result): any => {
                    const url = result['text'];
                    const hashindex = url.indexOf('/#/');
                    // could just whole page redirect to url
                    if (hashindex > 0) {
                        let regex = /[?&]([^=#]+)=([^&#]*)/g,
                            params = {},
                            match;
                        while ((match = regex.exec(url))) {
                            params[match[1]] = match[2];
                        }
                        this._router.navigate(
                            [url.split('/#')[1].split('?')[0]],
                            params
                        );
                    }
                })
                .catch((err) => {
                    console.error(err);
                });
        });
    }
}
