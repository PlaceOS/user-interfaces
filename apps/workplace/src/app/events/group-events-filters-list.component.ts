import { Component } from '@angular/core';
import { GroupEventsStateService } from './group-events-state.service';
import { differenceInDays, endOfDay, startOfDay } from 'date-fns';
import { map, take, tap } from 'rxjs/operators';

@Component({
    selector: 'group-events-filters-list',
    template: `
        <div
            class="my-2 w-[63rem] max-w-full mx-auto border border-base-300 p-4 bg-base-100 rounded"
        >
            <div class="flex items-center justify-between space-x-2 mb-4">
                <div>
                    {{ (this_period | async) ? 'This' : 'Upcoming' }}
                    {{ (period | async) === 'week' ? 'Week' : 'Month' }}
                </div>
            </div>
            <div class="flex flex-wrap">
                <div
                    class="flex items-center px-4 py-3 rounded-3xl border border-base-400 m-1 text-sm"
                >
                    {{ (options | async)?.date | date: 'MMM d, y' }}
                    &ndash;
                    {{ (options | async)?.end | date: 'MMM d, y' }}
                </div>
                <div
                    class="flex items-center pl-4 pr-1 rounded-3xl border border-base-400 m-1"
                    *ngFor="let tag of (filters | async)?.tags || []"
                >
                    <div class="mr-2 flex-1 text-sm">{{ tag }}</div>
                    <button icon matRipple (click)="removeTag(tag)">
                        <app-icon>close</app-icon>
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventsFiltersListComponent {
    public readonly options = this._state.options;
    public readonly filters = this._state.filters;
    public readonly this_period = this._state.options.pipe(
        map(({ date, end }) => {
            return (
                Date.now() >= startOfDay(date).valueOf() &&
                Date.now() < endOfDay(end || date).valueOf()
            );
        })
    );
    public readonly period = this._state.options.pipe(
        map(({ date, end }) =>
            Math.abs(differenceInDays(date, end || Date.now())) > 7
                ? 'month'
                : 'week'
        )
    );

    constructor(private _state: GroupEventsStateService) {}

    public async removeTag(tag: string) {
        const tags = (await this.filters.pipe(take(1)).toPromise())?.tags || [];
        this._state.setFilters({ tags: tags.filter((_) => _ !== tag) });
    }
}
