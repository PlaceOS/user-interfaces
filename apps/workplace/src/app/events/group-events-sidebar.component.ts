import { Component } from '@angular/core';
import { AsyncHandler, SettingsService } from '@placeos/common';
import {
    addDays,
    addMonths,
    endOfDay,
    format,
    startOfMonth,
    startOfWeek,
} from 'date-fns';
import { BehaviorSubject } from 'rxjs';
import { GroupEventsStateService } from './group-events-state.service';
import { take } from 'rxjs/operators';

@Component({
    selector: `group-events-sidebar`,
    template: `
        <div class="flex flex-col bg-base-100 w-[18rem] h-full">
            <div class="flex items-center space-x-2 p-2">
                <button
                    btn
                    matRipple
                    class="flex-1 rounded-3xl"
                    [class.inverse]="(period | async) !== 'week'"
                    (click)="period.next('week')"
                >
                    Week
                </button>
                <button
                    btn
                    matRipple
                    class="flex-1 rounded-3xl"
                    [class.inverse]="(period | async) !== 'month'"
                    (click)="period.next('month')"
                >
                    Month
                </button>
            </div>
            <div class="flex flex-col items-center space-y-2 pb-2 px-2">
                <mat-form-field
                    appearance="outline"
                    class="w-full no-subscript"
                >
                    <mat-select
                        [(ngModel)]="selected_range"
                        (ngModelChange)="setPeriod($event)"
                        placeholder="Select Period"
                    >
                        <mat-option
                            [value]="range.id"
                            *ngFor="let range of period_list"
                        >
                            {{ range.display }}
                        </mat-option>
                    </mat-select>
                </mat-form-field>
            </div>
            <hr class="border-base-200 w-[calc(100%-1rem)] mx-auto" />
            <date-calendar
                [ngModel]="(options | async).date"
                (ngModelChange)="setPeriodFromDate($event)"
            ></date-calendar>
            <hr class="border-base-200 w-[calc(100%-1rem)] mx-auto" />
            <div class="flex flex-col flex-1 overflow-auto">
                <h2 class="text-lg font-medium p-4">Filters</h2>
                <div
                    class="flex flex-col space-y-2 px-4"
                    *ngIf="(tags | async)?.length"
                >
                    <h3>Tags</h3>
                    <button
                        matRipple
                        class="flex items-center rounded w-full text-left"
                        *ngFor="let tag of tags | async"
                        (click)="toggleTag(tag)"
                    >
                        <mat-checkbox
                            [ngModel]="(filters | async)?.tags?.includes(tag)"
                        >
                            {{ tag }}
                        </mat-checkbox>
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventsSidebarComponent extends AsyncHandler {
    public period = new BehaviorSubject<'week' | 'month'>('week');
    public period_list = [];
    public selected_range: number;
    public readonly options = this._state.options;
    public readonly filters = this._state.filters;
    public readonly tags = this._state.tags;

    constructor(
        private _settings: SettingsService,
        private _state: GroupEventsStateService
    ) {
        super();
    }

    public ngOnInit(): void {
        this.subscription(
            'period',
            this.period.subscribe(() => {
                this._generatePeriods();
                if (this.period_list.length) {
                    this.setPeriod(this.period_list[0].id);
                    this.selected_range = this.period_list[0].id;
                }
            })
        );
        this._generatePeriods();
        if (this.period_list.length) {
            this.setPeriod(this.period_list[0].id);
            this.selected_range = this.period_list[0].id;
        }
    }

    public async toggleTag(tag: string) {
        const tags = (await this.filters.pipe(take(1)).toPromise())?.tags || [];
        if (tags.includes(tag)) {
            this._state.setFilters({ tags: tags.filter((_) => _ !== tag) });
        } else {
            this._state.setFilters({ tags: [...tags, tag] });
        }
    }

    public setPeriodFromDate(date: number) {
        for (const period of this.period_list) {
            if (date >= period.start && date <= period.end) {
                this.selected_range = period.id;
                this.setPeriod(period.id);
                return;
            }
        }
    }

    public setPeriod(id: string) {
        const { start, end } = this.period_list.find((_) => _.id === id);
        this._state.setOptions({ date: start, end });
    }

    private _generatePeriods() {
        const periods = [];
        const period_type = this.period.value;
        let date = Date.now();
        const end_date = addDays(date, 12 * 30).valueOf();
        const week_offset = this._settings.get('app.week_start') || 0;
        if (period_type === 'month') {
            date = startOfMonth(date).valueOf();
        } else if (period_type === 'week') {
            date = startOfWeek(date, { weekStartsOn: week_offset }).valueOf();
        }
        while (date < end_date) {
            if (period_type === 'week') {
                const end = endOfDay(addDays(date, 6)).valueOf();
                periods.push({
                    id: date,
                    start: date,
                    end,
                    display: `${format(
                        Math.max(Date.now(), date),
                        'EEE, do MMM'
                    )} â€“ ${format(end, 'do MMM')}`,
                });
                date = addDays(date, 7).valueOf();
            } else if (period_type === 'month') {
                const end = endOfDay(addDays(addMonths(date, 1), -1)).valueOf();
                periods.push({
                    id: date,
                    start: date,
                    end,
                    display: `${format(date, 'MMMM yyyy')}`,
                });
                date = addMonths(date, 1).valueOf();
            } else break;
        }
        this.period_list = periods;
    }
}
