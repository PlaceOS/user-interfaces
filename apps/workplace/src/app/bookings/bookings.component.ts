import { Component, OnInit } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

import { clearBookingFormData, clearDeskBookingFormData } from '@user-interfaces/bookings';
import { BaseClass, HashMap, SettingsService } from '@user-interfaces/common';


export interface BookingFlowEvent {
    /** Type of flow event */
    type: 'previous' | 'next' | 'reset' | 'close' | 'error';
    /** Step which the event was emitted from */
    step: string;
    /** Additional metadata for the event */
    metadata?: HashMap;
}

@Component({
    selector: 'app-bookings',
    templateUrl: './bookings.component.html',
    styleUrls: ['./bookings.component.scss'],
})
export class BookingsComponent extends BaseClass implements OnInit {
    /** Booking flow */
    public flow: 'spaces' | 'desks' | 'qr' | '';
    /** Whether to show the overlay menu */
    public show_menu: boolean;

    constructor(
        private _route: ActivatedRoute,
        private _router: Router,
        private _settings: SettingsService
    ) {
        super();
    }

    public ngOnInit(): void {
        this._settings.title = 'Book';
        this.subscription(
            'route.params',
            this._route.paramMap.subscribe((params) => {
                if (params.has('flow')) {
                    this.flow = params.get('flow') as any;
                } else {
                    this.flow = '';
                }
            })
        );
        this.subscription(
            'query.params',
            this._route.queryParamMap.subscribe((params) => {
                if (params.has('new')) {
                    clearDeskBookingFormData();
                    clearBookingFormData();
                    this._router.navigate([], {
                        queryParams: {
                            new: null,
                        },
                        replaceUrl: true,
                        queryParamsHandling: 'merge',
                    });
                }
            })
        );
    }

    public spaces() {
        this._router.navigate(['spaces'], { relativeTo: this._route });
    }
}
