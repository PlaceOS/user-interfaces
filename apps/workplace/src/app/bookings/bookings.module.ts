import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule, Route } from '@angular/router';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { BookingsComponent } from './bookings.component';
import { SharedContentModule } from '../ui/shared.module';
import { SharedExploreModule } from '@user-interfaces/explore';
import { SharedUsersModule } from '@user-interfaces/users';

import { QrReaderComponent } from './qr-reader/qr-reader.component';

import { SpaceFlowComponent } from './space-flow/space-flow.component';
import { SpaceFlowFormComponent } from './space-flow/space-flow-form.component';
import { SpaceFlowListingComponent } from './space-flow/space-flow-listing.component';
import { SpaceFlowConfirmComponent } from './space-flow/space-flow-confirm.component';

import { DeskFlowComponent } from './desk-flow/desk-flow.component';
import { DeskFlowListingComponent } from './desk-flow/desk-flow-listing.component';
import { DeskFlowConfirmModalComponent } from './desk-flow/desk-flow-confirm-modal.component';
import { DeskFlowQuestionsModalComponent } from './desk-flow/desk-flow-questions-modal.component';

const ROUTES: Route[] = [
    { path: ':flow', component: BookingsComponent },
    { path: ':flow/:step', component: BookingsComponent },
    { path: '', redirectTo: 'spaces' },
];

@NgModule({
    declarations: [
        BookingsComponent,

        DeskFlowComponent,
        DeskFlowListingComponent,
        DeskFlowConfirmModalComponent,
        DeskFlowQuestionsModalComponent,

        SpaceFlowComponent,
        SpaceFlowFormComponent,
        SpaceFlowListingComponent,
        SpaceFlowConfirmComponent,

        QrReaderComponent,
    ],
    imports: [
        CommonModule,
        RouterModule.forChild(ROUTES),
        FormsModule,
        ReactiveFormsModule,
        SharedContentModule,
        SharedUsersModule,
        SharedExploreModule
    ],
})
export class BookingsModule {

}
