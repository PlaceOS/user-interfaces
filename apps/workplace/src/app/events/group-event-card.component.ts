import { Component, Input } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { SettingsService } from '@placeos/common';
import { GroupEventDetailsModalComponent } from '@placeos/bookings';

@Component({
    selector: 'group-event-card',
    template: `
        <button
            matRipple
            (click)="viewDetails()"
            *ngIf="!featured; else featured_card"
            class="border border-base-300 hover:border-info flex flex-col bg-base-100 rounded-xl shadow hover:shadow-2xl overflow-hidden w-60 h-[20rem]"
        >
            <div
                class="flex items-center justify-between h-28 w-full bg-base-200 overflow-hidden"
            >
                <img
                    *ngIf="event.images?.length"
                    auth
                    [source]="event.images[0]"
                    class="object-cover"
                />
            </div>
            <div class="p-4 flex-1 h-1/2">
                <div class="opacity-60 text-sm text-left">
                    {{ event.date | date: 'EEE d MMM' }},
                    {{ event.date | date: time_format }}
                </div>
                <h2
                    class="text-xl mb-2 text-left truncate"
                    [title]="event.title"
                >
                    {{ event.title }}
                </h2>
                <div
                    class="opacity-60 text-xs flex-1 overflow-hidden h-[4.5rem] mb-2"
                >
                    {{ event.description }}
                </div>
                <div class="flex items-center space-x-2 text-sm">
                    <app-icon class="text-info">place</app-icon>
                    <div class="" *ngIf="item?.linked_event?.system_id">
                        {{
                            (item?.linked_event?.system_id | space | async)
                                ?.display_name
                        }}
                    </div>
                    <div
                        class="opacity-30"
                        *ngIf="!item?.linked_event?.system_id"
                    >
                        Remote event
                    </div>
                </div>
                <div class="flex items-center space-x-2 text-sm">
                    <app-icon class="text-info">people</app-icon>
                    <div class="">
                        {{ event.attendees?.length || '0' }} attending
                    </div>
                </div>
            </div>
        </button>
        <ng-template #featured_card>
            <button
                matRipple
                (click)="viewDetails()"
                class="border border-base-300 hover:border-info flex bg-base-100 rounded-xl shadow hover:shadow-2xl overflow-hidden w-[48rem] max-w-full h-56 mx-auto"
            >
                <div
                    class="flex items-center justify-between h-full w-1/2 max-w-[20rem] bg-base-200 overflow-hidden"
                >
                    <img
                        *ngIf="event.images?.length"
                        auth
                        [source]="event.images[0]"
                        class="object-cover"
                    />
                </div>
                <div
                    class="absolute top-0 left-0 rounded-br-xl py-2 pl-2 pr-4 space-x-2 bg-info text-info-content flex items-center text-sm"
                >
                    <app-icon class="text-base">star</app-icon>
                    <div class="uppercase">Featured</div>
                </div>
                <div details class="flex px-8 py-4 space-x-4">
                    <div class="flex flex-col items-center">
                        <div class="text-sm opacity-30 ">
                            {{ event.date | date: 'MMM' }}
                        </div>
                        <div class="text-lg">{{ event.date | date: 'd' }}</div>
                    </div>
                    <div class="flex flex-col space-y-2">
                        <h3 class="text-left">{{ event.title }}</h3>
                        <div time class="text-sm opacity-30">
                            {{ event.date | date: 'EEEE' }}
                            {{ event.date | date: time_format }} -
                            {{
                                event.date + event.duration * 60 * 1000
                                    | date: time_format
                            }}
                        </div>
                        <div class="h-20 overflow-hidden">
                            {{ event.description }}
                        </div>
                        <div class="flex items-center space-x-2 text-sm">
                            <app-icon class="text-info">place</app-icon>
                            <div class="" *ngIf="event.location">
                                {{ event.location }}
                            </div>
                            <div class="opacity-30" *ngIf="!event.location">
                                Remote event
                            </div>
                        </div>
                        <div class="flex items-center space-x-2 text-sm">
                            <app-icon class="text-info">people</app-icon>
                            <div class="">
                                {{ event.attendees?.length || '0' }} attending
                            </div>
                        </div>
                    </div>
                </div>
                <div
                    class="absolute top-4 right-4 bg-secondary text-secondary-content rounded px-4 py-2 w-32 text-center truncate"
                >
                    View Details
                </div>
            </button>
        </ng-template>
    `,
    styles: [
        `
            button {
                transition: box-shadow 300ms, border 200ms;
            }
        `,
    ],
})
export class GroupEventCardComponent {
    @Input() public event: any;
    @Input() public featured: boolean;

    public get time_format(): string {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _dialog: MatDialog
    ) {}

    public ngOnInit(): void {}

    public viewDetails(): void {
        console.log('View Details:', this.event);
        this._dialog.open(GroupEventDetailsModalComponent, {
            data: this.event,
        });
    }
}
