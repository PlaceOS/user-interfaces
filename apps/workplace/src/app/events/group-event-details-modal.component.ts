import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA } from '@angular/material/dialog';
import { Booking } from '@placeos/bookings';
import { SettingsService } from '@placeos/common';

@Component({
    selector: `group-event-details-modal`,
    template: `
        <div
            class="relative w-[48rem] max-w-[calc(100vw-1rem)] max-h-[80vh] overflow-hidden"
        >
            <div
                class="flex items-center justify-between h-52 w-full bg-base-200 overflow-hidden"
            >
                <img
                    *ngIf="booking.images?.length"
                    auth
                    [source]="event.images[0]"
                    class="object-cover"
                />
            </div>
            <div
                class="absolute top-0 left-0 rounded-br py-2 pl-2 pr-4 space-x-2 bg-info text-info-content flex items-center text-sm"
            >
                <app-icon class="text-base">star</app-icon>
                <div class="uppercase">Featured</div>
            </div>
            <button icon mat-dialog-close class="absolute top-1 right-1">
                <app-icon>close</app-icon>
            </button>
            <div
                class="flex items-center justify-between py-4 px-8 border-b border-base-200"
            >
                <h3 class="text-left text-xl">
                    {{ booking.title }}
                </h3>
                <div class="flex items-center space-x-2">
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content space-x-2"
                    >
                        <app-icon>star</app-icon>
                        <div class="pr-2">Interested</div>
                    </button>
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content space-x-2"
                    >
                        <app-icon>help</app-icon>
                        <div class="pr-2">Going</div>
                    </button>
                    <button
                        btn
                        matRipple
                        class="clear bg-base-200 text-base-content w-[2.75rem]"
                    >
                        <app-icon class="text-2xl">more_horiz</app-icon>
                    </button>
                </div>
            </div>
            <div class="flex overflow-y-auto overflow-x-hidden p-8">
                <div class="flex flex-col space-y-2">
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>person</app-icon>
                        </div>
                        <div>Event by {{ booking.user_name }}</div>
                    </div>
                    <h3 class="font-medium pt-4">When and where</h3>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>calendar_today</app-icon>
                        </div>
                        <div class="flex flex-col">
                            <div class="text-sm">Date and Time</div>
                            <div class="text-sm opacity-30">
                                {{ booking.date | date: 'EEEE, d MMMM, yyyy' }}
                                . {{ booking.date | date: time_format }} -
                                {{
                                    booking.date + booking.duration * 60 * 1000
                                        | date: time_format
                                }}
                            </div>
                        </div>
                    </div>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>place</app-icon>
                        </div>
                        <div class="text-sm">
                            {{ booking.location }}
                            <span *ngIf="!booking.location" class="opacity-30">
                                Remote Event
                            </span>
                        </div>
                    </div>
                    <div class="flex items-center space-x-4">
                        <div
                            class="flex items-center justify-center w-10 h-10 bg-base-200 rounded-full"
                        >
                            <app-icon>person</app-icon>
                        </div>
                        <div>
                            {{ attendance }} going,
                            {{ booking.attendees?.length - attendance }}
                            interested
                        </div>
                    </div>
                    <h3 class="font-medium pt-4">About this event</h3>
                    <div class="text-sm pb-4">
                        {{ booking.description }}
                        <span *ngIf="!booking.description" class="opacity-30">
                            No description
                        </span>
                    </div>
                </div>
                <div class="flex"></div>
            </div>
        </div>
    `,
    styles: [``],
})
export class GroupEventDetailsModalComponent {
    public booking: any = this._data;

    public get time_format() {
        return this._settings.time_format;
    }

    public get attendance() {
        return (
            this.booking.attendees?.filter((_) => _.status === 'accepted')
                ?.length || 0
        );
    }

    constructor(
        @Inject(MAT_DIALOG_DATA) private _data: Booking,
        private _settings: SettingsService
    ) {}

    public ngOnInit(): void {}
}
