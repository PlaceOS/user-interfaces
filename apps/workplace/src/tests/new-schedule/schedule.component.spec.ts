import { createRoutingFactory, SpectatorRouting } from '@ngneat/spectator/jest';
import { MockComponent, MockProvider } from 'ng-mocks';
import { FooterMenuComponent } from '../../app/components/footer-menu.component';
import { TopbarComponent } from '../../app/components/topbar.component';
import { ScheduleFiltersComponent } from '../../app/new-schedule/schedule-filters.component';
import { ScheduleSidebarComponent } from '../../app/new-schedule/schedule-sidebar.component';
import { ScheduleMobileCalendarComponent } from '../../app/new-schedule/schedule-mobile-calendar.component';
import { ScheduleComponent } from '../../app/new-schedule/schedule.component';
import { MatProgressBarModule } from '@angular/material/progress-bar';
import { ScheduleStateService } from '../../app/new-schedule/schedule-state.service';
import { BehaviorSubject } from 'rxjs';
import { EventCardComponent, EventFormService } from '@placeos/events';
import { BookingCardComponent } from '@placeos/bookings';
import { MatDialog } from '@angular/material/dialog';
import { FormsModule } from '@angular/forms';
import { SettingsService } from '@placeos/common';

describe('ScheduleComponent', () => {
    let spectator: SpectatorRouting<ScheduleComponent>;
    const createComponent = createRoutingFactory({
        component: ScheduleComponent,
        declarations: [
            MockComponent(ScheduleSidebarComponent),
            MockComponent(ScheduleMobileCalendarComponent),
            MockComponent(ScheduleFiltersComponent),
            MockComponent(EventCardComponent),
            MockComponent(BookingCardComponent),
            MockComponent(TopbarComponent),
            MockComponent(FooterMenuComponent),
        ],
        providers: [
            MockProvider(ScheduleStateService, {
                filtered_bookings: new BehaviorSubject([]),
                loading: new BehaviorSubject(false),
                date: new BehaviorSubject(0),
                toggleType: jest.fn(),
                setDate: jest.fn(),
            }),
            MockProvider(EventFormService, { newForm: jest.fn() }),
            MockProvider(MatDialog, { open: jest.fn() }),
            MockProvider(SettingsService, { get: jest.fn() }),
        ],
        imports: [MatProgressBarModule, FormsModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });
});
