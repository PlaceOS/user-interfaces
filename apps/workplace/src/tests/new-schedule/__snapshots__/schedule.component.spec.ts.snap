// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <topbar />
  <div
    class="flex flex-col sm:flex-row flex-1 h-1/2 bg-[#E5E5E5] relative"
  >
    <schedule-sidebar
      class="hidden sm:block"
    />
    <div
      class="w-full bg-white border-b border-gray-300 sm:hidden"
    >
      <schedule-mobile-calendar />
    </div>
    <div
      class="flex-1 h-full p-4 overflow-auto space-y-2"
    >
      <schedule-filters />
      <h3
        class="font-medium my-2"
      >
         Thu 01 Jan 1970 
        
      </h3>
      <div
        class="w-full p-8 flex flex-col items-center justify-center space-y-4"
      >
        <img
          class="mr-4"
          src="assets/img/no-events.svg"
        />
        <p
          class="opacity-30"
        >
           No events and bookings for Thursday, 01 Jan 1970 
        </p>
      </div>
      
    </div>
    
  </div>
  <footer-menu />
  
</div>
`;
